src/main/java/com/aws/greengrass/clientdevices/auth/api/DomainEvents.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.clientdevices.auth.api;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+
+@SuppressWarnings({"rawtypes","unchecked"})
+public class DomainEvents {
+    private final Map<Class, CopyOnWriteArrayList<DomainEventListener>> eventListeners = new ConcurrentHashMap<>();
+
+    @FunctionalInterface
+    public interface DomainEventListener<T extends DomainEvent> {
+        void handle(T event);
+    }
+
+    /**
+     * Register event listener.
+     * @param listener Event listener callback
+     * @param clazz    Type of domain event
+     * @param <T>      Type of domain event
+     */
+    public <T extends DomainEvent> void registerListener(DomainEventListener<T> listener, Class<T> clazz) {
+        CopyOnWriteArrayList<DomainEventListener> listeners =
+                eventListeners.computeIfAbsent(clazz, (k) -> new CopyOnWriteArrayList<>());
+        listeners.addIfAbsent(listener);
+    }
+
+    /**
+     * Emit an event.
+     * @param domainEvent Domain event
+     * @param <T>         Type of domain event
+     */
+    public <T extends DomainEvent> void emit(T domainEvent) {
+        List<DomainEventListener> listeners = eventListeners.get(domainEvent.getClass());
+        if (listeners != null) {
+            for (DomainEventListener<T> listener : listeners) {
+                listener.handle(domainEvent);
+            }
+        }
```

Comment:
Method name `emit()` sounds bit misleading here as it inherently handles the emitted event. May be rename to `trigger()`?

Reply:
How it's handled is an implementation detail. The caller does not know about what (if anything) happens. It is simply emitting an event